RFC36: remove header deps immature rule (#240)
nervosnetwork/rfcs: rfcs/0036-remove-header-deps-immature-rule/0036-remove-header-deps-immature-rule.md

---
Number: "0036"
Category: Consensus (Hard Fork)
Status: Draft
Author: Ian Yang
Organization: Nervos Foundation
Created: 2021-02-07
---

# Remove Header Deps Immature Rule

## Abstract

> This document proposes removing the loading header immature rule.

Author: Suggest to use the uniform term to refer the rule such as "Header Deps Immature Rule".

-- suggestion proposed by @chanhsu001 offline

In the consensus ckb2019, the header dep must reference the block which is 4 epochs ago. After this RFC is activated, the transaction can use any existing blocks in the chain as the header dep.

## Motivation

Header dep is a useful feature for dApps developers because script can use it to read the header of a block in the chain, or verify that an input cell or dep cell is in a specific block in the chain.

The immature rule prevents the usage of header deps in many scenarios because the script must reference the block about 16 hours ago.

> The intention of the immature rule is like the cellbase immature rule, a transaction with header deps and all its descendants can be invalidated after a chain reorganization [^1], because the referenced block may be rollbacked.

Author: I feel that the reading has come to an abrupt end here, only saying that the "Immature Rule" is necessary, but then it does not say why the rule should be abolished. Talking about the problem of chain reorganization, when the rule is abolished, how to solve?  

-- suggestion proposed by @chanhsu001 offline

---
Number: "0036"
Category: Consensus (Hard Fork)
Status: Draft
Author: Ian Yang
Organization: Nervos Foundation
Created: 2021-02-07
---

# Remove Header Deps Immature Rule

## Abstract

This document proposes removing the *[Loading Header Immature Rule]*.

[Loading Header Immature Rule]: ../0009-vm-syscalls/0009-vm-syscalls.md#loading-header-immature-error

In the consensus ckb2019, the header dep must reference the block which is 4 epochs ago. After this RFC is activated, the transaction can use any existing blocks in the chain as the header dep.

## Motivation

Header dep is a useful feature for dApps developers because script can use it to read the header of a block in the chain, or verify that an input cell or dep cell is in a specific block in the chain.

The *Loading Header Immature Rule* prevents the usage of header deps in many scenarios because the script must reference the block about 16 hours ago.

> The intention of the immature rule is like the cellbase immature rule, a transaction with header deps and all its descendants can be invalidated after a chain reorganization [^1], because the referenced block may be rolled back. Removing the rule lets dApps developers to trade off between responsive header reference and reliable transaction finality.

Reviewer: ```suggestion
The intention of the immature rule is like the cellbase immature rule. A transaction and all its descendants may be invalidated after a chain reorganization [^1], because its header deps referred to stale or orphan blocks. Removing the rule lets dApps developers trade-off between responsive header reference and reliable transaction finality.
```

[^1]: Chain reorganization happens when the node found a better chain with more accumulated proved work and it has to rollback blocks to switch to the new chain.

## Specification

> This RFC must be activated via a hard fork. After activation, the consensus no longer verifies that the referenced block in the header deps has been mined 4 epochs ago.

Reviewer 2: ```suggestion
This RFC must be activated via a hard fork. After activation, the consensus no longer verifies that the referenced block in the header deps is mined 4 epochs ago.
```
